src/core/components/files-regular/add-async-iterator.js

```diff
@@ -0,0 +1,200 @@
+const log = require('debug')('ipfs:add')
+log.error = require('debug')('ipfs:add:error')
+
+function noop () {}
+
+module.exports = function (self) {
+  // Internal add func that gets used by all add funcs
+  return async function * addAsyncIterator (source, options) {
+    options = options || {}
+
+    let chunkerOptions = parseChunkerString(options.chunker)
+
+    const opts = Object.assign({}, {
+      shardSplitThreshold: self._options.EXPERIMENTAL.sharding
+        ? 1000
+        : Infinity
+    }, options, {
+      chunker: chunkerOptions.chunker,
+      chunkerOptions: chunkerOptions.chunkerOptions
+    })
+
+    // CID v0 is for multihashes encoded with sha2-256
+    if (opts.hashAlg && opts.cidVersion !== 1) {
+      opts.cidVersion = 1
+    }
+
+    let total = 0
+
+    const prog = opts.progress || noop
+    const progress = (bytes) => {
+      total += bytes
+      prog(total)
+    }
+
+    opts.progress = progress
+
+    if (Buffer.isBuffer(source) || typeof source === 'string') {
+      source = [
+        source
+      ]
+    }
+
+    const iterator = pipe(
+      source,
+      validateInput(),
+      normalizeInput(opts),
+      doImport(self, opts),
+      prepareFile(self, opts),
+      preloadFile(self, opts),
+      pinFile(self, opts)
+      // sort?
+    )
+
+    for await (const thing of iterator) {
+      yield thing
+    }
+  }
+}
+
+function validateInput () {
+  return async function * (source) {
+    for await (const data of source) {
+      validateAddInput(data)
+
+      yield data
+    }
+  }
+}
+
+function normalizeContent (content) {
+  if (supportsFileReader && kindOf(content) === 'file') {
+    return streamFromFileReader(content)
+  }
+
+  // pull stream source
+  if (isSource(content)) {
+    return toAsyncIterator(content)
+  }
+
+  if (typeof content === 'string') {
+    return Buffer.from(content)
+  }
+
+  if (Array.isArray(content) && content.length && !Array.isArray(content[0])) {
+    return [content]
+  }
+
+  return content
+}
+
+function normalizeInput (opts) {
+  return async function * (source) {
+    for await (let data of source) {
+      if (data.content) {
+        data.content = normalizeContent(data.content)
+      } else {
+        data = {
+          path: '',
+          content: normalizeContent(data)
+        }
+      }
+
+      if (opts.wrapWithDirectory && !data.path) {
+        throw new Error('Must provide a path when wrapping with a directory')
+      }
+
+      yield data
+    }
+  }
+}
+
+function doImport (ipfs, opts) {
+  return function (source) {
+    return importer(source, ipfs._ipld, opts)
+  }
+}
+
+function prepareFile (ipfs, opts) {
+  return async function * (source) {
+    for await (const file of source) {
+      let cid = file.cid
+      const b58Hash = cid.toBaseEncodedString()
+
+      if (opts.onlyHash) {
+        yield {
+          path: file.path === undefined ? b58Hash : (file.path || ''),
+          hash: b58Hash,
+          size: file.unixfs.fileSize()
```

Comment:
This will be different to when you don't specify `onlyHash` as the node has been serialized so includes the protobuf overhead.  Typically we only show the user the hash when they've said `onlyHash` so I don't know how much of a problem this is.  I think we want to switch this over to only show file sizes eventually anyway so it might not be a big deal.